File: day1.R
library(tidyverse)
library(stringi)

input <- tibble(input = read_lines("data/1.txt")) 

strings = rev(c("one"   = "1",
                "two"   = "2",
                "three" = "3",
                "four"  = "4",
                "five"  = "5",
                "six"   = "6",
                "seven" = "7",
                "eight" = "8",
                "nine"  = "9"))

strings_rev <- strings
names(strings_rev) <- stri_reverse(names(strings))

solve <- function(input, r1 = "[0-9]", r2 = "[0-9]"){
  input |> 
    mutate(first_number = str_extract(input, r1)) |> 
    mutate(last_number = str_extract(stri_reverse(input), r2)) |> 
    mutate(across(first_number:last_number,
                  ~str_replace_all(., c(strings, strings_rev)))) |> 
    mutate(combined = as.integer(paste0(first_number, last_number))) |> 
    summarize(answer = sum(combined)) |> 
    pull()
}
  
input |> solve()
  
input |> solve(paste0("[0-9]|", paste(names(strings), 
                                      collapse = "|")),
               paste0("[0-9]|", paste(stri_reverse(names(strings)), 
                                      collapse = "|")))

File: day1.jl
d = readlines("data/1.txt")

∘̇(f,g) = x -> f(g.(x))

p1 = parse.(Int, (map(x -> x[1].captures[1] * x[end].captures[1], 
                           collect.(eachmatch.([r"([0-9])"], d))))) |> sum

regex_p2 = r"([0-9]|nine|eight|seven|six|five|four|three|two|one)"

strings =  Dict("one"   => "1",
                "two"   => "2",
                "three" => "3",
                "four"  => "4",
                "five"  => "5",
                "six"   => "6",
                "seven" => "7",
                "eight" => "8",
                "nine"  => "9")

p2 = parse.(Int, map(x -> replace(x[1].captures[1], strings...) *
                          replace(x[end].captures[1], strings...), 
              collect.(eachmatch.([regex_p2], d, overlap = true)))) |> sum
    